### PR TITLE
[MSHARED-1410] Track used declared dependency usage by referencing classes

### PR DESCRIPTION
**Pull request by jaredstehler**
As a follow-on to https://issues.apache.org/jira/browse/MSHARED-1407

This will also map uses of used declared dependencies to their referencing classes.

We use this in our internal dependency analyzer to support validation for migrating classes to new packages - for classes in a migration, we select their new artifact as a used dependency and not their old one.

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MSHARED) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes. Also be sure having selected the correct component.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MSHARED-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MSHARED-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the integration tests successfully (`mvn -Prun-its clean verify`).

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

File: src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java [hunk 1]
```diff
@@ -150,7 +155,7 @@ private static Set<Artifact> getTestArtifactsWithNonTestScope(Set<Artifact> test
         return nonTestScopeArtifacts;
     }
 
-    private Map<Artifact, Set<String>> buildArtifactClassMap(MavenProject project, ClassesPatterns excludedClasses)
+    protected Map<Artifact, Set<String>> buildArtifactClassMap(MavenProject project, ClassesPatterns excludedClasses)
```
Comment (jaredstehler) [inline src/main/java/org/apache/maven/shared/dependency/analyzer/DefaultProjectDependencyAnalyzer.java:158]: our internal dependency analyzer uses this method to compute used dependency classes which are provided by multiple artifacts

**Comment by jaredstehler** [comment]
gentle bump; anything I can do to help get this integrated?

**Review by slawekjaranowski — APPROVED** [review]

**Comment by slawekjaranowski** [comment]
@jaredstehler - thanks for patience

**Comment by jira-importer** [comment]
Resolve #237

**Pull request merged**